test/unit_tests/kinsol/CXX_serial/kin_test_getjac.cpp

```diff
@@ -0,0 +1,300 @@
+
+  // Create KINSOL memory structure
+  void* kin_mem = KINCreate(sunctx);
+  if (check_ptr(kin_mem, "KINCreate")) return 1;
+
+  int flag = KINInit(kin_mem, res, uu);
+  if (check_flag(flag, "KINInit")) return 1;
+
+  SUNMatrix A = SUNDenseMatrix(3, 3, sunctx);
+  if (check_ptr(A, "SUNDenseMatrix")) return 1;
+
+  SUNLinearSolver LS = SUNLinSol_Dense(uu, A, sunctx);
+  if (check_ptr(LS, "SUNLinSol_Dense")) return 1;
+
+  // Disable the linear solver setup function and attach custom solve function
+  LS->ops->setup = nullptr;
+  LS->ops->solve = DenseSetupAndSolve;
+
+  flag = KINSetLinearSolver(kin_mem, LS, A);
+  if (check_flag(flag, "KINSetLinearSolver")) return 1;
+
+  // Call main solver
+  flag = KINSol(kin_mem, uu, KIN_NONE, scale, scale);
+  if (check_flag(flag, "KINSol")) return 1;
+
+  long int nni;
+  flag = KINGetNumNonlinSolvIters(kin_mem, &nni);
+  if (check_flag(flag, "KINGetNumNonlinSolvIters")) return 1;
+
+  // Get the internal finite difference approximation to J
+  SUNMatrix Jdq;
+  flag = KINGetJac(kin_mem, &Jdq);
+  if (check_flag(flag, "KINGetJac")) return 1;
+
+  // Get the iteration the Jacobian was computed
+  long int nni_Jdq;
+  flag = KINGetJacNumIters(kin_mem, &nni_Jdq);
+  if (check_flag(flag, "KINGetJacNumSteps")) return 1;
+
+  // Compute the Jacobian at the returned solution
+  SUNMatrix Jtrue = SUNDenseMatrix(3, 3, sunctx);
+  if (check_ptr(Jtrue, "SUNDenseMatrix")) return 1;
+
+  flag = J(uu, nullptr, Jtrue, nullptr, nullptr, nullptr);
+  if (check_flag(flag, "J")) return 1;
+
+  // Compare finite difference and true Jacobian
+  sunrealtype* Jdq_data = SUNDenseMatrix_Data(Jdq);
+  if (check_ptr(Jdq_data, "SUNDenseMatrix_Data")) return 1;
+
+  sunrealtype* Jtrue_data = SUNDenseMatrix_Data(Jtrue);
+  if (check_ptr(Jtrue_data, "SUNDenseMatrix_Data")) return 1;
+
+  // Output Jacobian data
+  std::cout << std::scientific;
+  std::cout << std::setprecision(std::numeric_limits<sunrealtype>::digits10);
+  std::cout << "nni = " << nni << std::endl;
+  std::cout << "Jac nni = " << nni_Jdq << std::endl;
+  std::cout << std::endl;
+  std::cout << std::setw(8) << std::right << "Index" << std::setw(25)
+            << std::right << "J DQ" << std::setw(25) << std::right << "J true"
+            << std::setw(25) << std::right << "absolute difference"
+            << std::setw(25) << std::right << "relative difference" << std::endl;
+  for (int i = 0; i < 4 * 25 + 8; i++) std::cout << "-";
+  std::cout << std::endl;
+
+  sunindextype ldata = SUNDenseMatrix_LData(Jtrue);
+  for (sunindextype i = 0; i < ldata; i++)
+  {
+    std::cout << std::setw(8) << std::right << i << std::setw(25) << std::right
+              << Jdq_data[i] << std::setw(25) << std::right << Jtrue_data[i]
+              << std::setw(25) << std::right
+              << std::abs(Jdq_data[i] - Jtrue_data[i])
+              << std::setw(25) << std::right
+              << std::abs(Jdq_data[i] - Jtrue_data[i])/Jtrue_data[i]
+              << std::endl;
+  }
+
```

Comment:
If we could figure out why the results from this test show such large differences, then I would recommend that we compute a norm of the difference between `Jdq` and `Jtrue`, and use that to gauge pass/fail for the test in a precision-dependent manner.